### ISSUE TITLE
Dataset Optimization does not complete with a large number of Datasets

### ISSUE THREAD

**Issue by sseveran** [tensorflow/tensorflow#29753]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 1.13.1
- Python version: 3.6.7
- CUDA/cuDNN version: 10.1
- GPU model and memory: Geforce RTX 2080 ti

**Describe the current behavior**
I am trying to use tf.data.experimental.choose_from_datasets the create specifically constructed batches from a large number of datasets, ~7300. Each of the datasets I want to sample from is a TFRecord Dataset. It fails with a DeadlineExceededError (stack trace below).

The model is implemented as an estimator. I attempted to adjust the operation timeout but that does not seem to have any effect.

A few observations:

1. I know I may have pushed this a bit far. If I should actually build either a completely custom dataset or a custom operator that could take a stream of examples and build the batches with some sort of internal buffer let me know.
1. It is much slower (~5x) to create the datasets in Graph Mode as opposed to eager mode.
1. Eager mode has no problem yielding examples if I just loop over dataset and print them out.
1. The graph that is created by the estimator is extremely large (graph.pbtxt of 2.5GB)

**Describe the expected behavior**
It would be nice if this dataset could execute in graph mode.

**Code to reproduce the issue**
```python
def build_triplet_dataset(root, num_classes, classes_per_batch, num_examples, num_frames, repeat=None, num_gpus=1) -> tf.data.Dataset:
    cores = multiprocessing.cpu_count()

    def _generator():
        while True:
            result = []
            for _ in range(classes_per_batch):
                x = random.randint(0, num_classes)
                result.append([x] * num_examples)
            yield list(itertools.chain(*result))

    def _per_class_dataset(path) -> tf.data.Dataset:
        dataset = tf.data.TFRecordDataset(path, compression_type='GZIP')
        dataset = dataset.apply(tf.data.experimental.shuffle_and_repeat(num_gpus * num_examples, repeat))
        dataset = dataset.map(lambda x: _parse_example(x, num_frames), num_parallel_calls=cores)
        dataset = dataset.apply(tf.data.experimental.unbatch())
        dataset = dataset.shuffle(num_gpus*num_examples)
        return dataset

    files = glob.glob(os.path.join(root, "*.tfrecord.gz"))
    print('Loading Datasets')
    datasets = list(tqdm(map(lambda x: _per_class_dataset(x), files)))

    choice_dataset = tf.data.Dataset.from_generator(lambda: _generator(), (tf.int64),
                                                    output_shapes=(tf.TensorShape([num_examples * classes_per_batch])))
    choice_dataset = choice_dataset.apply(tf.data.experimental.unbatch())

    dataset = tf.data.experimental.choose_from_datasets(datasets, choice_dataset)
    dataset = dataset.batch(classes_per_batch * num_examples)
    dataset = dataset.prefetch(tf.data.experimental.AUTOTUNE)
    return dataset
```

**Other info / logs**
Caused by op 'OptimizeDataset', defined at:
  File "sonia/models/my_model/train.py", line 14, in <module>
    ds.train_and_evaluate(model_params, hyperparameters, [], [])
  File "/home/steve/source/core/python/sonia/models/deep_speaker/model2.py", line 453, in train_and_evaluate
    steps=None))
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/training.py", line 471, in train_and_evaluate
    return executor.run()
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/training.py", line 611, in run
    return self.run_local()
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/training.py", line 712, in run_local
    saving_listeners=saving_listeners)
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 358, in train
    loss = self._train_model(input_fn, hooks, saving_listeners)
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 1122, in _train_model
    return self._train_model_distributed(input_fn, hooks, saving_listeners)
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 1185, in _train_model_distributed
    self._config._train_distribute, input_fn, hooks, saving_listeners)
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 1206, in _actual_train_model_distributed
    input_fn, model_fn_lib.ModeKeys.TRAIN, strategy)
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 985, in _get_iterator_from_input_fn
    iterator = result.make_initializable_iterator()
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow/python/distribute/values.py", line 1193, in make_initializable_iterator
    self._dataset, self._devices)
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow/python/data/ops/multi_device_iterator_ops.py", line 180, in __init__
    self._dataset._as_variant_tensor(),  # pylint: disable=protected-access
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 2993, in _as_variant_tensor
    self._input_dataset._as_variant_tensor(),  # pylint: disable=protected-access
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 3013, in _as_variant_tensor
    **flat_structure(self))
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow/python/ops/gen_dataset_ops.py", line 3036, in optimize_dataset
    output_shapes=output_shapes, name=name)
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3300, in create_op
    op_def=op_def)
  File "/home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1801, in __init__
    self._traceback = tf_stack.extract_stack()

DeadlineExceededError (see above for traceback): meta_optimizer exceeded deadline.
         [[node OptimizeDataset (defined at /home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py:985) ]]
         [[node MultiDeviceIteratorInit (defined at /home/steve/venv/sonia/lib/python3.6/site-packages/tensorflow_estimator/python/estimator/estimator.py:985) ]]

**Comment by sseveran** [comment]
One brief update. I disabled optimization by setting options.experimental_optimization.apply_default_optimizations = False which leads to

INFO:tensorflow:Initialize strategy                                                                                                                                                                                
2019-06-15 13:38:17.149375: I tensorflow/stream_executor/dso_loader.cc:152] successfully opened CUDA library libcublas.so.10.0 locally
terminate called after throwing an instance of 'std::system_error'                                                                                                                                                 
  what():  Resource temporarily unavailable                                                                                                                                                                        
Aborted

**Comment by ravikyram** [comment]
Please provide us complete code to reproduce the issue.Thanks

**Comment by ravikyram** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
